### PR TITLE
Migration script to single chart upgrade.

### PR DESCRIPTION
**Pull request by diptiranjanpx**
Signed off : Diptiranjan
  
migration.sh will be used for upgrading from multi charts (1.2.*) to single chart (1.3.0).
The script has all the steps automated to do the helm upgrade.

### CODE REVIEW

File: single_chart_migration/migration.sh [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+    $helm_cmd get values $pxmonitor_release --namespace $namespace -o yaml >> helm_values.yaml
+fi
+if [ "$pxls_enabled" != "" ]; then
+    $helm_cmd get values $pxls_release --namespace $namespace -o yaml >> helm_values.yaml
+fi
+
+# Change the annotations
+echo "Step-4 : Changing annotations of px-monitor and px-license-server componets if they have been installed"
+if [ "$pxmonitor_enabled" != "" ]; then
+    echo "Changing the annotations of existing px-monitor componenets"
```
Comment (sharma-tapas) [inline single_chart_migration/migration.sh:199]: components

File: single_chart_migration/migration.sh [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    change_annotation "deployment" "$deploymentListMonitor"
+    change_annotation "prometheus" "$prometheusMonitor"
+    change_annotation "prometheusrule" "$prometheusruleMonitor"
+    change_annotation "servicemonitor" "$servicemonitorListMonitor"
+    delete_resource "statefulset" "$statefulsetListMonitorDelete"
+    delete_resource "deployment" "$deploymentListMonitorDelete"
+fi
+
+if [ "$pxls_enabled" != "" ]; then
+    echo "Changing the annotations of existing px-license-server componenets"
```
Comment (sharma-tapas) [inline single_chart_migration/migration.sh:219]: components

**Review by sharma-tapas — COMMENTED** [review]
Please run a spell check on the echo messages, rest looks good.

**Comment by sharma-tapas** [comment]
@diptiranjanpx Also add a Readme.md file to this directory.

**Review by sharma-tapas — COMMENTED** [review]
Add Readme.md file to this directory, it should have
1. How to download the script
2. How to run the script
3. How to get the parameters needed to run the script.

File: single_chart_migration/migration.sh [hunk 3]
```diff
@@ -0,0 +1,258 @@
+#!/bin/bash
+
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply single_chart_migration/migration.sh:2]: Added the following.

File: single_chart_migration/migration.sh [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+cortex_nginx_deployment="pxcentral-cortex-nginx"
+ls_configmap="pxcentral-ls-configmap"
+post_install_job="pxcentral-post-install-hook"
+
+usage()
+{
+   echo ""
+   echo "Usage: $0 --namespace <namespace> --helmrepo <helm repo name> --kubeconfig <kubeconfig file name> "
+   echo -e "\t--namespace <namespace> namespace in which px-central charts are installed"
+   echo -e "\t--helmrepo <helm repo name for px-central components> helm repo name , can get with helm repo list command"
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply single_chart_migration/migration.sh:55]: Made the optional args explicit. Any other way you wanted?

File: single_chart_migration/migration.sh [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+            echo "kubeconfig file = $2"
+            kubeconfig=$2
+            shift
+            shift
+            ;;
+    esac
+done
+
+if [ "$namespace" == "" ] || [ "$helmrepo" == "" ]; then
+    usage
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply single_chart_migration/migration.sh:194]: Done

File: single_chart_migration/migration.sh [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+    exit 1
+fi
+
+if [ "$kubeconfig" != "" ]; then
+    kubectl_cmd="kubectl --kubeconfig $kubeconfig"
+    helm_cmd="helm --kubeconfig $kubeconfig"
+fi
+
+echo "Step-1 : Finding out the enabled modules"
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply single_chart_migration/migration.sh]: Done.

File: single_chart_migration/migration.sh [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+set_enabled_modules $namespace
+echo "pxbackup: $pxbackup_enabled, pxmonitor: $pxmonitor_enabled, pxlicenseserver: $pxls_enabled"
+
+echo "Step-2 : Finding the installed releases for the components"
+find_releases $namespace
+echo "pxbackup release: $pxbackup_release, pxmonitor release: $pxmonitor_release, pxlicenseserver release: $pxls_release"
+
+# Get all values in a single file
+echo "Step-3 : Putting the values in helm_values.yaml file"
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply single_chart_migration/migration.sh]: done.

File: single_chart_migration/migration.sh [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+if [ "$kubeconfig" != "" ]; then
+    kubectl_cmd="kubectl --kubeconfig $kubeconfig"
+    helm_cmd="helm --kubeconfig $kubeconfig"
+fi
+
+echo "Step-1 : Finding out the enabled modules"
+set_enabled_modules $namespace
+echo "pxbackup: $pxbackup_enabled, pxmonitor: $pxmonitor_enabled, pxlicenseserver: $pxls_enabled"
+
+echo "Step-2 : Finding the installed releases for the components"
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply single_chart_migration/migration.sh]: Made it "Searching the installed helm releases for the components" as we are checking for helm releases with helm list command.

File: single_chart_migration/migration.sh [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+fi
+
+# Delete the post-install job
+echo "Step-5 : Deleting post-install hook job before upgrade"
+delete_job_cmd="$kubectl_cmd --namespace $namespace delete job $post_install_job"
+echo $delete_job_cmd
+$delete_job_cmd
+
+# Do the upgrade
+echo "Step-6 : Going to do upgrade now"
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply single_chart_migration/migration.sh]: Done

File: single_chart_migration/migration.sh [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+$helm_cmd get values $pxbackup_release --namespace $namespace -o yaml > helm_values.yaml
+if [ "$pxmonitor_enabled" == true ]; then
+    $helm_cmd get values $pxmonitor_release --namespace $namespace -o yaml >> helm_values.yaml
+fi
+if [ "$pxls_enabled" != "" ]; then
+    $helm_cmd get values $pxls_release --namespace $namespace -o yaml >> helm_values.yaml
+fi
+
+# Change the annotations
+echo "Step-4 : Changing annotations of px-monitor and px-license-server componets if they have been installed"
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply single_chart_migration/migration.sh]: Done

File: single_chart_migration/migration.sh (hunk 1 above)
Comment (diptiranjanpx) [reply single_chart_migration/migration.sh]: Done

File: single_chart_migration/migration.sh (hunk 2 above)
Comment (diptiranjanpx) [reply single_chart_migration/migration.sh]: Done

File: single_chart_migration/README.md [hunk 11]
```diff
@@ -0,0 +1,39 @@
+# Upgrading to Single Chart in 1.3.0
```
Comment (sharma-tapas) [inline single_chart_migration/README.md:1]: Upgrade from 1.2.X to 1.3.0

File: single_chart_migration/README.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+```console
+$ helm search repo portworx
+```
+
+- Run migration.sh to upgrade to 1.3.0
+
+```console
+$ ./migration.sh --namespace <namespace> --helmrepo <helm repo name>
+```
```
Comment (sharma-tapas) [inline single_chart_migration/README.md:39]: Add a section to get the details such as 
1. How to get namespace of the install
2. How to check the current installed components
3. Should we add the new helm repo as the same name or a new name

**Review by sharma-tapas — COMMENTED** [review]
Mark review comments as resolved once fixed, add a little more details of the script to the Readme.md

**Review by sharma-tapas — APPROVED** [review]
lgtm

**Pull request merged**